File: main/symmetric-tree/symmetric-tree.cpp
/*
Structure of the node of the tree is as
struct Node
{
	int key;
	struct Node* left, *right;
};
*/

namespace {
    bool isSymmetric(const struct Node* const left,
                     const struct Node* const right)
    {
        if (left == nullptr) return right == nullptr;
        
        if (right == nullptr || left->key != right->key) return false;
        
        return isSymmetric(left->left, right->right)
                && isSymmetric(left->right, right->left);
    }
}

// complete this function
// return true/false if the is Symmetric or not
bool isSymmetric(struct Node* const root)
{
    return root == nullptr || isSymmetric(root->left, root->right);
}

File: main/symmetric-tree/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    @staticmethod
    def isSymmetric(root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return root is None or Solution._mirrors(root.left, root.right)

    @staticmethod
    def _mirrors(left, right):
        if left is None or right is None:
            return left is None and right is None

        return (left.val == right.val
                and Solution._mirrors(left.left, right.right)
                and Solution._mirrors(left.right, right.left))
